File: java/Secure.java
/**
 * Secure Java code samples for SAST false positive testing
 * These are SAFE implementations that should NOT be flagged
 * Author: Test Suite
 * Created: 2025-01-21
 */

import java.sql.*;
import java.io.*;
import java.nio.file.*;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.util.Base64;

public class Secure {
    
    // ✅ SAFE: Parameterized SQL query (should NOT be flagged)
    public ResultSet getUserData(String userId) throws SQLException {
        String query = "SELECT * FROM users WHERE id = ?";
        PreparedStatement stmt = connection.prepareStatement(query);
        stmt.setString(1, userId); // Parameterized query
        return stmt.executeQuery();
    }
    
    // ✅ SAFE: Parameterized SQL with LIKE (should NOT be flagged)
    public ResultSet searchUsers(String searchTerm) throws SQLException {
        String query = "SELECT * FROM users WHERE name LIKE ?";
        PreparedStatement stmt = connection.prepareStatement(query);
        stmt.setString(1, "%" + searchTerm + "%"); // Parameterized
        return stmt.executeQuery();
    }
    
    // ✅ SAFE: Command execution with whitelist (should NOT be flagged)
    public void executeCommand(String command) throws IOException {
        String[] allowedCommands = {"ls", "pwd", "date"};
        boolean isAllowed = false;
        for (String allowed : allowedCommands) {
            if (command.equals(allowed)) {
                isAllowed = true;
                break;
            }
        }
        if (isAllowed) {
            Runtime.getRuntime().exec(command);
        }
    }
    
    // ✅ SAFE: Secrets from environment variables (should NOT be flagged)
    private static final String API_KEY = System.getenv("API_KEY");
    private static final String PASSWORD = System.getenv("DB_PASSWORD");
    private static final String SECRET = System.getenv("SECRET_KEY");
    
    // ✅ SAFE: Path validation before file access (should NOT be flagged)
    public String readFile(String filename) throws IOException {
        Path basePath = Paths.get("/var/www/data");
        Path filePath = basePath.resolve(filename).normalize();
        if (!filePath.startsWith(basePath)) {
            throw new SecurityException("Invalid path");
        }
        return new String(Files.readAllBytes(filePath));
    }
    
    // ✅ SAFE: Strong crypto - SHA-256 (should NOT be flagged)
    public String hashPassword(String password) throws Exception {
        MessageDigest md = MessageDigest.getInstance("SHA-256");
        byte[] hash = md.digest(password.getBytes());
        return Base64.getEncoder().encodeToString(hash);
    }
    
    // ✅ SAFE: Secure random (should NOT be flagged)
    public String generateToken() {
        SecureRandom random = new SecureRandom();
        byte[] bytes = new byte[32];
        random.nextBytes(bytes);
        return Base64.getEncoder().encodeToString(bytes);
    }
    
    // ✅ SAFE: XSS prevention with encoding (should NOT be flagged)
    public void displayMessage(String message) {
        String escaped = message
            .replace("&", "&amp;")
            .replace("<", "&lt;")
            .replace(">", "&gt;")
            .replace("\"", "&quot;")
            .replace("'", "&#x27;");
        response.getWriter().println("<div>" + escaped + "</div>");
    }
}

File: kotlin/Secure.kt
/**
 * Secure Kotlin code samples for SAST false positive testing
 * These are SAFE implementations that should NOT be flagged
 * Author: Test Suite
 * Created: 2025-01-21
 */

import java.security.MessageDigest
import java.security.SecureRandom
import java.io.File
import java.sql.PreparedStatement
import java.util.Base64

// ✅ SAFE: Parameterized SQL query (should NOT be flagged)
fun getUserData(userId: String): ResultSet {
    val query = "SELECT * FROM users WHERE id = ?"
    val stmt: PreparedStatement = connection.prepareStatement(query)
    stmt.setString(1, userId) // Parameterized query
    return stmt.executeQuery()
}

// ✅ SAFE: Parameterized SQL with LIKE (should NOT be flagged)
fun searchUsers(searchTerm: String): ResultSet {
    val query = "SELECT * FROM users WHERE name LIKE ?"
    val stmt: PreparedStatement = connection.prepareStatement(query)
    stmt.setString(1, "%$searchTerm%") // Parameterized
    return stmt.executeQuery()
}

// ✅ SAFE: Command execution with whitelist (should NOT be flagged)
fun executeCommand(command: String) {
    val allowedCommands = listOf("ls", "pwd", "date")
    if (command in allowedCommands) {
        Runtime.getRuntime().exec(command)
    }
}

// ✅ SAFE: Secrets from environment variables (should NOT be flagged)
val API_KEY = System.getenv("API_KEY")
val PASSWORD = System.getenv("DB_PASSWORD")
val SECRET = System.getenv("SECRET_KEY")

// ✅ SAFE: Path validation before file access (should NOT be flagged)
fun readFile(filename: String): String {
    val basePath = java.io.File("/var/www/data")
    val filePath = java.io.File(basePath, java.io.File(filename).name)
    
    val canonicalBase = basePath.canonicalFile
    val canonicalFile = filePath.canonicalFile
    
    if (!canonicalFile.path.startsWith(canonicalBase.path)) {
        throw SecurityException("Invalid path")
    }
    
    return canonicalFile.readText()
}

// ✅ SAFE: Strong crypto - SHA-256 (should NOT be flagged)
fun hashPassword(password: String): String {
    val md = MessageDigest.getInstance("SHA-256")
    val hash = md.digest(password.toByteArray())
    return Base64.getEncoder().encodeToString(hash)
}

// ✅ SAFE: Secure random (should NOT be flagged)
fun generateToken(): String {
    val random = SecureRandom()
    val bytes = ByteArray(32)
    random.nextBytes(bytes)
    return Base64.getEncoder().encodeToString(bytes)
}

// ✅ SAFE: XSS prevention with encoding (should NOT be flagged)
fun displayMessage(message: String) {
    val escaped = message
        .replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
        .replace("\"", "&quot;")
        .replace("'", "&#x27;")
    response.writer.println("<div>$escaped</div>")
}

// ✅ SAFE: Safe deserialization (should NOT be flagged)
fun deserializeData(data: ByteArray): Any? {
    // In production, use safe deserialization libraries
    // This is a placeholder showing the pattern
    return null // Safe - no unsafe deserialization
}

